admiral/pkg/clusters/virtualservice_routing.go

```diff
@@ -0,0 +1,628 @@
+	}
+
+	// Here we will create a separate canary destination for the canary FQDN
+	// This is needed to provide users to validate their canary endpoints
+	serviceNamespace := serviceInstance.Namespace
+	canaryServiceName := rollout.Spec.Strategy.Canary.CanaryService
+	host := canaryServiceName + common.Sep + serviceNamespace + common.GetLocalDomainSuffix()
+	if destinations[canaryFQDN] == nil {
+		destinations[canaryFQDN] = make([]*networkingV1Alpha3.RouteDestination, 0)
+	}
+	destinations[canaryFQDN] = append(destinations[canaryFQDN], getRouteDestination(host, meshPort, 0))
+
+	return nil
+}
+
+// addUpdateVirtualServicesForSourceIngress adds or updates the cross-cluster routing VirtualServices
+// This is where the VirtualServices are created using the services that were discovered during the
+// discovery phase.
+func addUpdateVirtualServicesForSourceIngress(
+	ctx context.Context,
+	ctxLogger *log.Entry,
+	remoteRegistry *RemoteRegistry,
+	sourceClusterToDestinations map[string]map[string][]*networkingV1Alpha3.RouteDestination) error {
+
+	if remoteRegistry == nil {
+		return fmt.Errorf("remoteRegistry is nil")
+	}
+
+	if len(sourceClusterToDestinations) == 0 {
+		return fmt.Errorf("no route destination found for the ingress virtualservice")
+	}
+
+	for sourceCluster, destination := range sourceClusterToDestinations {
+		rc := remoteRegistry.GetRemoteController(sourceCluster)
+
+		if rc == nil {
+			ctxLogger.Warnf(common.CtxLogFormat, "addUpdateVirtualServicesForSourceIngress",
+				"", "", sourceCluster, "remote controller not initialized on this cluster")
+			continue
+		}
+
+		virtualService, err := getBaseVirtualServiceForIngress()
+		if err != nil {
+			return err
+		}
+
+		vsName := ""
+		vsHosts := make([]string, 0)
+		tlsRoutes := make([]*networkingV1Alpha3.TLSRoute, 0)
+
+		for globalFQDN, routeDestinations := range destination {
+			hostWithoutSNIPrefix, err := getFQDNFromSNIHost(globalFQDN)
+			if err != nil {
+				ctxLogger.Warnf(common.CtxLogFormat, "addUpdateVirtualServicesForSourceIngress",
+					"", "", sourceCluster, err.Error())
+				continue
+			}
+			if !strings.HasPrefix(hostWithoutSNIPrefix, common.BlueGreenRolloutPreviewPrefix) &&
+				!strings.HasPrefix(hostWithoutSNIPrefix, common.CanaryRolloutCanaryPrefix) {
+				vsName = hostWithoutSNIPrefix + "-routing-vs"
+			}
+			if routeDestinations == nil || len(routeDestinations) == 0 {
+				ctxLogger.Warnf(common.CtxLogFormat, "addUpdateVirtualServicesForSourceIngress",
+					"", "", sourceCluster,
+					fmt.Sprintf("skipped adding host %s, no valid route destinaton found", hostWithoutSNIPrefix))
+				continue
+			}
+			tlsRoute := networkingV1Alpha3.TLSRoute{
+				Match: []*networkingV1Alpha3.TLSMatchAttributes{
+					{
+						Port:     common.DefaultMtlsPort,
+						SniHosts: []string{globalFQDN},
+					},
+				},
+				Route: routeDestinations,
+			}
+			tlsRoutes = append(tlsRoutes, &tlsRoute)
+			vsHosts = append(vsHosts, globalFQDN)
```

Comment:
Little confused when will we have multiple hosts ?

Reply:
is canary and default part of the same VS ?

Reply:
yes, they are. Do you see any issues with that?